src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.ReferenceHandlers.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.VS.References.Roslyn;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.References
+{
+    internal partial class ReferenceCleanupService
+    {
+        private abstract class ReferenceHandler
+        {
+            private readonly ProjectSystemReferenceType _referenceType;
+            private readonly string _schema;
+            private readonly string _referenceName;
+            private readonly string _itemSpecification;
+
+            private IProjectRuleSnapshot? _snapshotReferences;
+
+            protected ReferenceHandler(ProjectSystemReferenceType referenceType, string schema, string referenceName, string itemSpecification)
+            {
+                _referenceType = referenceType;
+                _schema = schema;
+                _referenceName = referenceName;
+                _itemSpecification = itemSpecification;
+            }
+
+            public abstract Task<bool> RemoveReferenceAsync(ConfiguredProject configuredProject, ProjectSystemReferenceInfo reference);
+
+            public IProjectRuleSnapshot GetProjectSnapshot(ConfiguredProject selectedConfiguredProject)
+            {
+                IProjectSubscriptionService? serviceSubscription = selectedConfiguredProject.Services.ProjectSubscription;
+                Assumes.Present(serviceSubscription);
+
+                serviceSubscription.ProjectRuleSource.SourceBlock.TryReceive(filter, out IProjectVersionedValue<IProjectSubscriptionUpdate> item);
+
+                _snapshotReferences = item.Value.CurrentState[_schema];
+
+                return _snapshotReferences;
+            }
+
+            private static bool filter(IProjectVersionedValue<IProjectSubscriptionUpdate> obj)
```

Comment:
Methods should start with an uppercase letter. You might also consider making this a lambda or local function.

Reply:
done.